### PR TITLE
Enhancement/Code Editor Instructor View: UI improvements

### PR DESCRIPTION
**Pull request by thilo-behnke**
### Checklist
- [ ] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] I documented my source code using the JavaDoc / JSDoc style.
- [x] ~I added integration test cases for the server (Spring) related to the features~
- [x] I added integration test cases for the client (Jest) related to the features
- [x] I added screenshots/screencast of my UI changes
- [x] I translated all the newly inserted strings

### Motivation and Context

Improve the instructor/ta screens for programming-exercises so that they are easier to use.

### Description

**Code-Editor:**
- Feature: Add resizing option for build output.
- Enhancement: Remove checkmark for indicating that file loading is done in ace editor.
- Styling: Make action buttons in file-browser lighter.
- Styling: Make vertical/horizontal resizing work as expected.
- Styling: Structure editor styles into component scss modules (before that there was a global file).

**Code-Editor-Instructor-View:**
- Enhancement: Don't reload the screen on first repo change (There was a route for not selecting a participation, this caused a visible reload of the page when a repository is selected the first time).
- Bugfix: Don't add the participation to the participation websocket cache (This could cause issues on other pages as the course-exercise-detail).

**Editable-Instructions:**
- Feature: Move the save problem statement button into the editable programming exercise component. This way also in the update/dialog programming exercise screen the problem statement can be saved invidividually.
- Feature: Move status icons into bottom area of editable instructions (similiar to file-browser).
- Enhancement: Hide status & save button if programming exercise is being created in update screen
- Bugfix: Also hide the update oracle structure tooltip, when the button is hidden.
- Styling: Fix command button styling in markdown editor (weird linebreaks on smaller devices).
- Styling: Increase default size of editable instructions
- Styling: Structure programming-exercise instruction styles into component scss modules (before that there was a global file).

**Markdown-Editor:**
- Feature: Add fullscreen mode to markdown editor
- Styling: Structure markdown-editor instruction styles into component a scss module. This way the markdown-editor adjusts its size now to the parent component.

:warning: Known styling issues in the code-editor :warning: 
1) In the instructor code editor, the height of the main area jumps a little, when switching from editor to preview in the instructions. This issue is very hard to pin down, I wouldn't invest more time into it personally.
2) In the code editor on very small screen when adjusting the width of the instructions it will overflow the main container to the right. This issue is also hard to solve (part of the max min width calculation needs to be done in JS because of InteractJs) and the best compromise to a resizable layout. I don't think it's a major issue because students can still just collapse the instructions.

### Steps for Testing

Every Feature/Enhancement/Bugfix has its own testing steps (see description above).
Most of the changes concern styling.

All components that are concerned by the changes:
- Code-Editor Student
- Code-Editor Instructor
- Programming-Exercise update/dialog/detail
- Markdown-Editor (and all components that use the markdown editor, only styling wise however)

### Screenshots

Code-Editor-Instructor before:

image

Code-Editor-Instructor after:

image

Markdown-Fullscreen mode:

image

### CODE REVIEW

**Comment by thilo-behnke** [comment]
The production js bundle of this branch ran into an issue with webpack, it leads to an error when being interpreted.
I'm looking into it.

File: src/main/webapp/i18n/de/editor.json [hunk 1]
```diff
@@ -6,17 +6,26 @@
             },
             "title": "Code Editor für Übung:",
             "instructions": "Aufgabenstellung &nbsp;",
-            "unsavedChanges": "Nicht gespeicherte Änderungen vorhanden.",
+            "unsavedChanges": "Ungespeichert.",
+            "unsavedChangesTooltip": "Nicht gespeicherte Änderungen vorhanden.",
             "savingChanges": "Änderungen werden gespeichert.",
+            "savingChangesTooltip": "Speichert...",
```
Comment (lovaiible) [inline src/main/webapp/i18n/de/editor.json]: Shouldn't the text for savingChanges and savingChangesTooltip be switched? :)

File: src/main/webapp/i18n/en/editor.json [hunk 2]
```diff
@@ -6,17 +6,26 @@
             },
             "title": "Code Editor for Exercise:",
             "instructions": "&nbsp; Instructions &nbsp;",
-            "unsavedChanges": "There are unsaved changes.",
+            "unsavedChanges": "Unsaved.",
+            "unsavedChangesTooltip": "There are unsaved changes.",
             "savingChanges": "Saving changes.",
+            "savingChangesTooltip": "Saving...",
```
Comment (lovaiible) [inline src/main/webapp/i18n/en/editor.json]: Shouldn't the text for savingChanges and savingChangesTooltip be switched? :)

File: src/main/webapp/app/entities/programming-exercise/instructions/programming-exercise-instruction-testcase-status.component.html [hunk 3]
```diff
@@ -1,6 +1,11 @@
-<span *ngIf="invalidTestCases.length + missingTestCases.length > 0">
-    <fa-icon icon="exclamation-triangle" size="2x" class="text-warning" [ngbTooltip]="tooltipContent"></fa-icon>
-</span>
+<div *ngIf="invalidTestCases.length + missingTestCases.length > 0">
+    <fa-icon icon="exclamation-triangle" class="text-warning mr-1" [ngbTooltip]="tooltipContent"></fa-icon>
+    <span class="text-info" jhiTranslate="artemisApp.programmingExercise.testCaseAnalysis.issues">Test case issues.</span>
+</div>
+<div *ngIf="invalidTestCases.length + missingTestCases.length <= 0">
+    <fa-icon icon="check-circle" class="text-success mr-1" [ngbTooltip]="tooltipContent"></fa-icon>
+    <span class="text-info" jhiTranslate="artemisApp.programmingExercise.testCaseAnalysis.ok">Test case issues.</span>
```
[the remark this answers is not in the document]
Comment (thilo-behnke) [reply src/main/webapp/app/entities/programming-exercise/instructions/programming-exercise-instruction-testcase-status.component.html]: Good catch, didn't notice that.

File: src/main/webapp/app/entities/programming-exercise/instructions/programming-exercise-editable-instruction.component.html [hunk 4]
```diff
@@ -1,17 +1,47 @@
...
     <jhi-markdown-editor
-        [defaultHeight]="markdownEditorHeight"
+        class="editable-instruction-container__editor"
         [domainCommands]="domainCommands"
-        [enableResize]="true"
         [markdown]="exercise.problemStatement"
         (markdownChange)="updateProblemStatement($event)"
     >
-        <jhi-programming-exercise-instructions #preview [participation]="participation" [exercise]="exercise"></jhi-programming-exercise-instructions>
+        <jhi-programming-exercise-instructions class="editable-instruction-container__instructions" #preview [participation]="participation" [exercise]="exercise">
```
[the remark this answers is not in the document]
Comment (thilo-behnke) [reply src/main/webapp/app/entities/programming-exercise/instructions/programming-exercise-editable-instruction.component.html]: You are right.

File: src/main/webapp/app/code-editor/build-output/code-editor-build-output.scss [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/*!* ==========================================================================
+Build-output
+========================================================================== *!*/
+
+.build-output {
+    height: inherit;
+
+    .build-output__content {
```
[the remark this answers is not in the document]
Comment (thilo-behnke) [reply src/main/webapp/app/code-editor/build-output/code-editor-build-output.scss]: You are right.

File: src/main/webapp/i18n/de/editor.json (hunk 1 above)
Comment (thilo-behnke) [reply src/main/webapp/i18n/de/editor.json]: Correct :)

File: src/main/webapp/i18n/en/editor.json (hunk 2 above)
Comment (thilo-behnke) [reply src/main/webapp/i18n/en/editor.json]: Correct :)

**Comment by thilo-behnke** [comment]
@krusche 
We checked the markdown-editor/fullscreen mode on safari with the following result:

1. The small overlap of the tab buttons comes from the height of the markdown editor content to be 100%. This is not a problem on Chrome, but for some reason on Safari. There was no easy fix without breaking something else...
2. When entering the fullscreen mode on Safari, the width now adjusts, but the height doesn't. The behavior is strange, sometimes opening the dev console seemed to help. We tried a couple of things but couldn't change it unfortunately.

**Comment by krusche** [comment]
please fix the merge conflicts

**Comment by thilo-behnke** [comment]
:warning: 
This was a really tough merge... I need to test this tomorrow morning again on our server.

**Comment by thilo-behnke** [comment]
@krusche I tested this and didn't notice any issues. Seems good!

**Pull request merged**